File: 739.daily-temperatures.py
#
# @lc app=leetcode id=739 lang=python3
#
# [739] Daily Temperatures
#


# @lc code=start
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        # maintain a stack, and the temperature inside only gets lower or equal
        # store (tempature, index) inside
        # when parse a new tempature, try push it into stack
        # pop and process if the top one is cooler
        stack = []
        n = len(temperatures)
        ans = [0] * n
        for idx, temp in enumerate(temperatures):
            # push
            while stack and stack[-1][0] < temp:
                _, targetIdx = stack.pop()
                ans[targetIdx] = idx - targetIdx
            stack.append((temp, idx))
        return ans


# @lc code=end

File: 739.daily-temperatures.go
/*
 * @lc app=leetcode id=739 lang=golang
 *
 * [739] Daily Temperatures
 */

// @lc code=start
package main

func dailyTemperatures(temperatures []int) []int {
	/*
	   Maintain a stack, store [val, idx]
	   Only decreasing,
	   iterate thru the arr, when the temp higher than top of the stack
	   keep popping, and update ans array
	*/
	stack := [][]int{}
	n := len(temperatures)
	ans := make([]int, n)
	for i, temp := range temperatures {
		for len(stack) > 0 && stack[len(stack)-1][0] < temp {
			idx := stack[len(stack)-1][1]
			ans[idx] = i - idx
			// pop
			stack = stack[:len(stack)-1]
		}
		stack = append(stack, []int{temp, i})
	}
	return ans
}

// @lc code=end
